KD_Lib/KD/text/Bert2Lstm/bert2lstm.py

```diff
@@ -0,0 +1,217 @@
+    :param optimizer_teacher (torch.optim.*): Optimizer used for training teacher. If None, AdamW is used.
+    :param loss_fn (torch.nn.module): Loss function
+    :param temp (float): Temperature parameter for distillation
+    :param distil_weight (float): Weight paramter for distillation loss
+    :param device (str): Device used for training; 'cpu' for cpu and 'cuda' for gpu
+    :param log (bool): True if logging required
+    :param logdir (str): Directory for storing logs
+    """
+
+    def __init__(
+        self,
+        student_model,
+        distill_train_loader,
+        distill_val_loader,
+        train_df,
+        val_df,
+        num_classes=2,
+        seed=42,
+        loss_fn=nn.MSELoss(),
+        temp=20.0,
+        distil_weight=0.5,
+        device="cpu",
+        log=False,
+        logdir="./Experiments",
+    ):
+
+        self.set_seed(42)
+
+        self.train_df, self.val_df = train_df, val_df
+
+        teacher_model = BertForSequenceClassification.from_pretrained(
+            "bert-base-uncased",
+            num_labels=num_classes,
+            output_attentions=False,
+            output_hidden_states=False,
+        )
+
+        optimizer_teacher = AdamW(teacher_model.parameters(), lr=2e-5, eps=1e-8)
+
+        self.bert_tokenizer = BertTokenizer.from_pretrained(
+            "bert-base-uncased", do_lower_case=True
+        )
+
+        super(Bert2LSTM, self).__init__(
+            teacher_model,
+            student_model,
+            distill_train_loader,
+            distill_val_loader,
+            optimizer_teacher,
+            None,
+            loss_fn,
+            temp,
+            distil_weight,
+            device,
+            log,
+            logdir,
+        )
+
+        self.optimizer_student = torch.optim.Adam(
+            self.student_model.parameters(), lr=2e-4
+        )
+
+    def set_seed(self, seed):
+        random.seed(seed)
+        np.random.seed(seed)
+        torch.manual_seed(seed)
+        torch.cuda.manual_seed_all(seed)
+
+    def _get_teacher_dataloaders(self, max_seq_length=128, batch_size=16, mode="train"):
+        """
+        Helper function for generating dataloaders for the teacher
+        """
+        df = self.val_df if (mode == "validate") else self.train_df
+
+        return get_bert_dataloader(
+            df, self.bert_tokenizer, max_seq_length, batch_size, mode
+        )
+
+    def calculate_kd_loss(self, y_pred_student, y_pred_teacher, y_true):
+        """
+        Function used for calculating the KD loss during distillation
+        :param y_pred_student (torch.FloatTensor): Prediction made by the student model
+        :param y_pred_teacher (torch.FloatTensor): Prediction made by the teacher model 
+        :param y_true (torch.FloatTensor): Original label
+        """
+
+        # soft_teacher_out = F.softmax(y_pred_teacher / self.temp, dim=0)
+        # soft_student_out = F.log_softmax(y_pred_student / self.temp, dim=0)
```

Comment:
why are these two commented out? In the original loss function we take T-softened probablity

Reply:
I must have commented those lines out while testing. I will uncomment those

Reply:
cool